dxr/app.py

```diff
@@ -315,36 +341,113 @@ def sidebar_links(sections):
+        lines = [doc['content'] for doc in line_docs]
+        if not contents:
+            # If contents are not provided, we can reconstruct them by
+            # stitching the lines together.
+            contents = ''.join(lines)
+        offsets = cumulative_sum(imap(len, lines))
+        # Construct skimmer objects for all enabled plugins that define a
+        # file_to_skim class.
+        skimmers = [plugin.file_to_skim(path,
+                                        contents,
+                                        name,
+                                        config.trees[tree],
+                                        file_doc,
+                                        line_docs,
+                                        current_app.vcs_caches[tree])
+                    for name, plugin in all_plugins().iteritems()
+                    if plugin in config.trees[tree].enabled_plugins
+                    and plugin.file_to_skim]
+        skim_links, refses, regionses, annotationses = skim_file(skimmers, len(line_docs))
+        index_refs = triples_from_es_refs(doc.get('refs', []) for doc in line_docs)
+        index_regions = triples_from_es_regions(doc.get('regions', []) for doc in line_docs)
+        tags = finished_tags(lines,
+                             chain(chain.from_iterable(refses), index_refs),
+                             chain(chain.from_iterable(regionses), index_regions))
         return render_template(
             'text_file.html',
-            **merge(common, file_vars, {
+            **merge(common, {
                 # Someday, it would be great to stream this and not concretize
                 # the whole thing in RAM. The template will have to quit
                 # looping through the whole thing 3 times.
-                'lines': [(html_line(doc['content'][0], doc.get('tags', [])),
-                           doc.get('annotations', [])) for doc in lines],
+                'lines': [(html_line(doc['content'], tags_in_line, offset),
+                           doc.get('annotations', []) + skim_annotations)
+                          for doc, tags_in_line, offset, skim_annotations
+                              in izip(line_docs, tags_per_line(tags), offsets, annotationses)],
                 'is_text': True,
-                'sections': sidebar_links(links)}))
+                'sections': sidebar_links(links + skim_links)}))
 
+@dxr_blueprint.route('/<tree>/rev/<revision>/<path:path>')
+def rev(tree, revision, path):
+    """Display a page showing the file at path at specified revision by
+    obtaining the contents from version control.
+    """
+    config = current_app.dxr_config
+    tree_config = config.trees[tree]
+    abs_path = join(tree_config.source_folder, path)
+    vcs = current_app.vcs_caches[tree].vcs_for_path(path)
+    if vcs:
+        contents = vcs.get_contents(relpath(abs_path, vcs.get_root_dir()), revision)
+        contents = contents.decode(tree_config.source_encoding)
```

Comment:
Rather than be inconsistent (by just catching any decode exceptions) and thereby showing binary files that *happen* to decode cleanly, I suppose we should run the contents through is_text().